HF21 OXEN -> SESH transition

tewinget opened the pull request.

jagerman committed:
WIP oxen->sent transition code

tewinget committed:
HF21 transition code mostly finished

the bulk of the code for transitioning to HF21 is present, and much
testing code is present.  This code does not yet work, and the tests are
not yet complete.

add hf21 transition mechanism to sqlite db

tewinget committed:
Post-hf21 purge and deactivation limit

After HF21, any nodes which did not transition to SENT stakes need to be
removed.  Doing them all at once would be chaotic with swarms, so they
will be removed by the existing decommission/deregister system.

That system now has a config-determined limit of deactivations per block
so that this transition (but also huge network outages) won't cause
swarms to disappear (or other possible chaos resulting from mass
removal).

tewinget committed:
Final HF20 block payout

Staked oxen addresses which are not registered for the SENT transition
will have their pending service node rewards paid out in full on the
final block before HF21.

Modified local devnet testing to leave a staked static-startup address
out of the registration for testing this (and adjusted expected numbers
accordingly).

tewinget committed:
correct removal of nodes which did not go past hf21

nodes which did not transition to hf21 will be removed by the
decommission->deregister mechanism rather than by the
bls-key-not-in-the-contract mechanism

tewinget committed:
HF21 fixes and debug logging

tewinget committed:
rough-but-complete script for hf21 transition info

script needs parameters set as appropriate:
  - oxen -> eth address mapping
  - transition bonus mapping
  - staking requirement and scaling ratio

TODO: some mainnet nodes have a larger stake, scaling needs to be normalized for those

tewinget committed:
change localdev initial staking req to match contract

tewinget committed:
testnet hf21 height

tewinget committed:
change testnet netid for chain restart

tewinget committed:
Don't apply HF dereg grace period to zombie nodes

They cannot ever come back, so there's really no point of leaving them
for a week.

(We *don't* want to immediately dereg them because going into decomm
still allows SN comms between storage servers, allowing the leaving
nodes to still transfer their files out to the remaining nodes upon
swarm dissolution).

tewinget committed:
make hf21 zombies not payable

nodes which do not transition at hardfork 21 should not be
in the reward payment rotation

tewinget committed:
remove erroneous hf21 transition change

tewinget committed:
add hf21 dry run rpc method

calling this method returns a before and after as if hf21 would
happen the very next block.  Because outstanding oxen rewards will
differ from the exact fork height (and from block to block), this
can only be approximate, but it should be fairly close to what the
actual result will be.

tewinget committed:
local devnet retry on json-rpc 'error'

tewinget committed:
hf21 transition: always set bls key if present, even if node fails transition

tewinget committed:
devnet restart for hf21 testing

tewinget committed:
add devnet reset old block hashes

tewinget committed:
devnet temp hf heights, double staking req

higher staking requirement to test hf21 transition correctly scaling stake
proportions in multi-contrib

tewinget committed:
fix inverted condition in daemon command

tewinget committed:
devnet timings underflow fix (and slow it down a little)

tewinget committed:
devnet hf20 height

tewinget committed:
(reverted later) devnet hf21 height and stop chain before

one line of this commit was kept and moved, making actually reverting it
and demolishing this commit tedious, so it stays, but the important
aspect of it which was to be reverted was later reverted.

tewinget committed:
redo/fix pre-hf21 bulk payout for non-registrants

The block before HF21 all addresses with accrued rewards which are
not registered for the HF21 transition with an ETH address will have
their OXEN balance paid out at that height rather than converted.

This was not working correctly for multiple reasons, but is now.

tewinget committed:
push devnet hf21 back a little

tewinget committed:
devnet hf21 parameters

tewinget committed:
fix uptime proofs using hf rules one block too early

tewinget committed:
stop using old monero-style keys altogether at HF21

tewinget committed:
fix stake scaling for hf21 transition for over-funded nodes

tewinget committed:
fix hf21_info stake scaling for over-funded nodes

also update devnet hf21 parameters

tewinget committed:
change devnet l2 tracker parameters

tewinget committed:
remove devnet stall-before-hf21 code

tewinget committed:
devnet hf dereg grace period 100 blocks

tewinget committed:
devnet reboot for testing mainnet test contracts

tewinget committed:
cleanup

tewinget committed:
clang format

Doy-lee commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -756,6 +766,15 @@ bool core::init(
+    if (height > 0) {
+        auto hf21_height = hard_fork_begins(m_nettype, hf::hf21_eth);
+        if (hf21_height && height > *hf21_height) {
+            service_node_list.while_locked([this]() { init_service_keys(true); });
```
I got mildly confused, didn't realise that `init_service_keys` is called twice and that the function handles that gracefully. Thought that there was a bug where it would just overwrite the Monero key with a runtime Ed25519 key that wasn't loaded off disk. 

Can we pull out the fixup of the monero key into a static freestanding function and reuse that in the `init_service_keys` section where we fixup the key and in the 2 places here where we convert on the hardforking block? 

Makes the logic of what's happening to the keys more explicit.

jagerman replied:
> Can we pull out the fixup of the monero key into a static freestanding function and reuse that in the init_service_keys section where we fixup the key and in the 2 places here where we convert on the hardforking block?

Tom and I talked about this a few days ago -- this is some very hacky code that can be removed (along with the entire Monero key loading) as soon as HF20 has happened, and so yeah, it's gross, but it's also getting reverted entirely in a few weeks.

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -3665,7 +3741,7 @@ void service_node_list::process_block(
 
     m_state.update_from_block(
-            blockchain.db(),
+            blockchain,
```
We should just use the new function prototype here and pass in `blockchain.db()` and `blockchain.maybe_sqlite_db()` since the call-site had to be updated anyway in this case to take `blockchain` instead of `blockchain.db`.

Which means then we can get rid of the helper function that takes the `Blockchain` and thunks into a function that pulls out the SQL DB. The SNL header is quite unwieldy so it'd be good to try and get rid of stuff from there where we can.

tewinget replied:
Good point, done.

Doy-lee reviewed (commented):
Generally LGTM, minor nitpicks.

Are we missing some code to block someone from requesting to leave the network after HF20? I think there might be an edge case we don't handle if we let people exit and their `requested_unlock_height` is set on Oxen but the equivalent mechanism is not set on the smart contract.

I couldn't come up with any other edge cases so far, seems like a bunch of the edge cases are already covered.

jagerman commented:
> Are we missing some code to block someone from requesting to leave the network after HF20? I think there might be an edge case we don't handle if we let people exit and their requested_unlock_height is set on Oxen but the equivalent mechanism is not set on the smart contract.

I don't think we need such a block.  Whenever an oxen unlock is processed, we set the exit height at +15d.  (That may well be from an unlock issued during HF19 if the HF20 window is less than 15d).  If that unlock height extends into HF21, I think it's fine: we will end up with a node that the contract does not know initiated an unlock, but that oxend does, and so when that unlock happens on the oxend-side, service nodes will start issuing voluntary exit signatures, and start issuing liquidation signatures 7d after that.

Before [this reversion](https://github.com/oxen-io/eth-sn-contracts/commit/a43793c3d3d42d6a06d3e345675db129df9c7113) that would have been a problem, but with that check gone the contract will happily exit anything the oxen network has signed off on.

tewinget committed:
hf21 ensure 'zombies' do not linger in recently removed

nodes which fail hf21 transition but were seeded into the contract need
to be removeable from the contract.  the ones which were *not* seeded
into the contract would not be removable from `recently_removed_nodes`
so should not be added there to begin with

Doy-lee commented:
LGTM

tewinget committed:
boost 1.87 deprecated removal fixes

Boost 1.87 removed many things which were previously only deprecated,
some of which we were still using.

tewinget committed:
remove boost version guard on typedef

changing from including boost asio io_service to io_context is roughly equivalent,
but the former has a typedef for compatibility.  I replicated the typedef,
but gated it on the boost version that removed the io_service include.

tewinget committed:
bump static build boost to 1.87

tewinget committed:
clang format to appease the lint gods

tewinget committed:
undo superfluous (invalid) boost-related change

tewinget committed:
boost is so much fun

tewinget committed:
thank you again, clang format!

tewinget committed:
fix sent transition circular dependency
